draft-irtf-cfrg-opaque.md

```diff
@@ -1338,70 +1173,182 @@ client_mac
 : An authentication tag computed over the handshake transcript computed using
 Km2, defined below.
 
-#### OPAQUE-3DH Functions {#opaque-3dh-functions}
+### Key Schedule Functions
 
-##### OPAQUE-3DH Message Protocol functions
+#### Transcript Functions
+
+The OPAQUE-3DH key derivation procedures make use of the functions below, re-purposed
+from TLS 1.3 {{?RFC8446}}.
 
 ~~~
-Start(credential_request, client_info)
+Expand-Label(Secret, Label, Context, Length) =
+    Expand(Secret, CustomLabel, Length)
+~~~
+
+Where CustomLabel is specified as:
+
+~~~
+struct {
+   uint16 length = Length;
+   opaque label<8..255> = "OPAQUE-" + Label;
+   opaque context<0..255> = Context;
+} CustomLabel;
+
+Derive-Secret(Secret, Label, Transcript-Hash) =
+    Expand-Label(Secret, Label, Transcript-Hash, Nx)
+~~~
+
+Note that the Label parameter is not a NULL-terminated string.
+
+The OPAQUE-3DH key schedule requires a preamble, which is computed as follows.
+
+~~~
+Preamble(client_identity, ke1, server_identity, inner_ke2)
+
+Input:
+- client_identity,  the optional encoded client identity, which is set to its public key if not specified
+- KE1, a KE1 message structure
+- server_identity, the optional encoded server identity, which is set to its public key if not specified
+- inner_ke2, a inner_ke2 structure as defined in KE2
+
+Output:
+- preamble, the protocol transcript with identities and messages
+
+Steps:
+1. preamble = concat("3DH",
+                     I2OSP(len(client_identity), 2), client_identity,
+                     KE1,
+                     I2OSP(len(server_identity), 2), server_identity,
+                     KE2.inner_ke2)
+2. Output preamble
+~~~
+
+#### Shared Secret Derivation
+
+The OPAQUE-3DH shared secret derive during the key exchange protocol is computed
+using the following function.
+
+~~~
+TripleDHIKM(sk1, pk1, sk2, pk2, sk3, pk3)
+
+Input:
+- skx, scalar to be multiplied with their corresponding pkx
+- pkx, element to be multiplied with their corresponding skx
+
+Output:
+- ikm, input key material
+
+Steps:
+1. dh1 = sk1 * pk1
+2. dh2 = sk2 * pk2
+3. dh3 = sk3 * pk3
+4. Output concat(dh1, dh2, dh3)
+~~~
+
+Using this shared secret, further keys used for encryption and authentication are
+computed using the following function.
+
+~~~
+DeriveKeys(ikm, preamble)
 
 Parameters:
-- Nn, the nonce length
+- Nx, the output size of the Extract function
```

Comment:
This function doesn't use Nx directly. Is it necessary specifying it here?

Reply:
Nope -- removed.